### PR TITLE
feat: supported Gemini Pro Vision

### LINKED ISSUE #35: Support Gemini Pro Vision model

### PR DESCRIPTION
**Pull request by adubovik**

### CODE REVIEW

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: failed

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: failed

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by dspashynskyi** [comment]
/deploy-review
>Deployment status: success

File: aidial_adapter_vertexai/chat_completion.py [hunk 1]
```diff
@@ -28,26 +30,33 @@ def __init__(self, region: str, project_id: str):
...
             prompt, discarded_messages_count = await model.truncate_prompt(
                 prompt, params.max_prompt_tokens
             )
 
         async def generate_response(usage: TokenUsage, choice_idx: int) -> None:
             with response.create_choice() as choice:
                 consumer = ChoiceConsumer(choice)
-                await model.chat(params, consumer, prompt)
-                usage.accumulate(consumer.usage)
+                if isinstance(prompt, UserError):
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/chat_completion.py]: I think it's better to fail fast (after line 42) and avoid further checks in the code.

File: aidial_adapter_vertexai/universal_api/errors.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+from aidial_sdk.chat_completion import Choice, Status
+from aiohttp import hdrs
+
+from aidial_adapter_vertexai.llm.exceptions import UserError
+
+
+def report_user_error(
+    choice: Choice, headers: Mapping[str, str], error: UserError
+) -> None:
+    is_chat_usage = headers.get(hdrs.AUTHORIZATION) is not None
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/universal_api/errors.py]: I'm not sure if this is reliable. Maybe the adapter could always do both things?

File: aidial_adapter_vertexai/universal_api/storage.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            ) as response:
+                response.raise_for_status()
+                meta = await response.json()
+                logger.debug(f"Uploaded file: url={url}, metadata={meta}")
+                return meta
+
+    async def upload_file_as_base64(
+        self, data: str, content_type: str
+    ) -> FileMetadata:
+        filename = _compute_hash_digest(data)
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/universal_api/storage.py:86]: Is it intentional that file will have same name for same data? What is expected when same data is uploaded twice?

File: aidial_adapter_vertexai/universal_api/storage.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    return base64.b64encode(bytes).decode("ascii")
+
+
+def _compute_hash_digest(file_content: str) -> str:
+    return hashlib.sha256(file_content.encode()).hexdigest()
+
+
+def _get_extension(content_type: str) -> Optional[str]:
+    if content_type.startswith("image/"):
+        return "." + content_type[len("image/") :]
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/universal_api/storage.py]: Maybe this one would be safer to use?
https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_extension

Comment (adubovik) [reply aidial_adapter_vertexai/universal_api/storage.py]: Indeed. Fixed.

File: aidial_adapter_vertexai/universal_api/storage.py (hunk 3 above)
Comment (adubovik) [reply aidial_adapter_vertexai/universal_api/storage.py:86]: Yes. Store is implemented via PUT request, which is supposed to be idempotent: two identical requests must result in the identical responses.

File: aidial_adapter_vertexai/universal_api/errors.py (hunk 2 above)
Comment (adubovik) [reply aidial_adapter_vertexai/universal_api/errors.py]: Yes, indeed, I like this.

File: aidial_adapter_vertexai/chat_completion.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_vertexai/chat_completion.py]: Fixed

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Review by Oleksii-Klimov — DISMISSED** [review]

**Review by Oleksii-Klimov — DISMISSED** [review]

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Comment by adubovik** [comment]
/deploy-review
>Deployment status: success

**Review by Oleksii-Klimov — APPROVED** [review]

**Pull request merged**